core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -687,4 +691,31 @@ public void postIndex(Engine.Index index, Throwable ex) {
         assertTrue(postIndexWithExceptionCalled.get());
     }
 
+    public void testOptimizeOnClosedShard() throws IOException, InterruptedException {
```

Comment:
I'm not sure whether this test buys us much, since we already test this on the engine level.. remove?